docs/javascript-typescript-interop.md

```diff
@@ -0,0 +1,98 @@
+
+1. Register a JavaScript function by calling `Blazor.registerFunction('functionName', functionImplementation)` in JavaScript. The following example registers two functions that echo a message. The `echo` function is synchronous, while the `echoAsync` version is asynchronous:
+
+    ```javascript
+    Blazor.registerFunction(
+      'echo',
+      function(message){
+        return message;
+      });
+
+    Blazor.registerFunction(
+      'echoAsync',
+      function(message){
+        return Promise.Resolve(message);
+      });
+    ```
+
+1. To invoke a JavaScript function from C#, use the `T Invoke<T>(functionName, args)` (synchronous) or `Task<T> InvokeAsync<T>(functionName, args)` (asynchronous) method from the `RegisteredFunction` class. The following example calls the two preceding JavaScript functions from C# code by indicating their function names and passing `message` arguments:
+
+    ```csharp
+    var helloWorld = RegisteredFunction.Invoke<string>("echo", "Hello world!");
+    var helloWorldAsync = 
+        await RegisteredFunction
+            .InvokeAsync<string>("echoAsync", "Hello world async!");
+    ```
+
+## Invoke .NET methods from JavaScript functions
+
+JavaScript code in the browser might be required to call .NET methods. For example, a .NET method call can be made when a callback in JavaScript is triggered.
+
+A .NET method call can be achieved by using the JavaScript functions `Blazor.invokeDotNetMethod` and `Blazor.invokeDotNetMethodAsync`. These two functions allow JavaScript to call synchronous and asynchronous .NET methods, respectively. The .NET method must have the following characteristics:
+
+* Static
+* Non-generic
+* Not overloaded
+* Concrete type method parameters
+* Parameters deserializable using JSON
+
+To call a .NET method, create the method in C# and invoke the method from JavaScript:
+
+1. Create the .NET method in C#. The following example creates a `TimeoutCallback` method that writes a message to the console when the method is invoked:
+
+    ```csharp
+    namespace Alerts
+    {
+        public class Timeout
+        {
+            public static void TimeoutCallback()
+            {
+                Console.WriteLine('Timeout triggered!');
+            }
+        }
+    }
+    ```
+
+1. Invoke the .NET method from JavaScript with a call to `Blazor.invokeDotNetMethod` (or `Blazor.invokeDotNetMethodAsync`). Specify the .NET assembly, namespace with class, and method name. The following example calls the `TimeoutCallback` .NET method in the preceding step:
+
+    ```javascript
+    Blazor.invokeDotNetMethod({
+      type: {
+        assembly: 'MyTimeoutAssembly',
+        name: 'Alerts.Timeout'
+      },
+      method: {
+        name: 'TimeoutCallback'
+      }
+    });
+    ```
+
+## Share interop code in a Blazor class library
+
+JavaScript interop code can be included in a Blazor class library (`dotnet new blazorlib`), which allows you to share the code in a NuGet package.
```

Comment:
I think we should discuss how the Blazor class library handles embedded the JS resources in the build assembly for you. You just put all your JS files under wwwroot and the build takes care of the rest. You can then reference the built package like you would any normal NuGet package and call into JS as if it were C#.